### PR TITLE
Provide a simple string matcher as the dual of the simple string interpolator

### PR DESCRIPTION
**Pull request by lihaoyi**
Examples:

```scala
val s"Hello, $name" = "Hello, James"
println(name)  // "James"
```

```scala
val s"$greeting, $name" = "Hello, James"
println(greeting)  // "Hello"
println(name)  // "James"
```

```scala
val TimeSplitter = "([0-9]+)[.:]([0-9]+)".r
val s"The time is ${TimeSplitter(hours, mins)}" = "The time is 10.50"
println(hours) // 10
println(mins) // 50
```

This is slightly random, but I think it fits very well as a counterpart
to the existing `s"..."` interpolator, has a very simple definition, has
lived in the ammonite-ops library as the `r` interpolator for a long
time, and I have generally found it very useful.

It supports simple globbing matches, without any fancy regex
group-matching or whatever. This is a nice fit for the simple semantics
of the existing `s` interpolator.

One subtlety is that in the case of ambiguity, the implementation gives
the characters to later matches rather than earlier matches:

```scala
scala> val s"$a-$b-$c-$d" = "1-2-3-4-5-6-7-8-9"
a: String = 1
b: String = 2
c: String = 3
d: String = 4-5-6-7-8-9
```

This is somewhat arbitrary. This is for simple matches; for anything
complex or ambiguous, you can use a full regex.

Fancier regex matching, e.g. using regex-syntax in the literal parts of the string or regex capturing groups to the bound variables, is out of scope for this PR. Maybe it can be done separately, but even if it does appear I think there's value in a simple matcher that people can use without worrying about needing to escape their `.`s and other regex special-characters.

Notably, this *cannot* be implemented in user-land right now as an extension method on `StringContext`, because the presence of the `s` method stops the extension method from being picked up.

What do people think? If it seems plausible that we could get this into the standard library, I'll be happy to add a few unit tests and do whatever polish is necessary.

### CODE REVIEW

File: src/library/scala/StringContext.scala [hunk 1]
```diff
@@ -92,7 +92,48 @@ case class StringContext(parts: String*) {
...
+     *    val TimeSplitter = "([0-9]+)[.:]([0-9]+)".r
+     *    val s"The time is ${TimeSplitter(hours, mins)}" = "The time is 10.50"
+     *    println(hours) // 10
+     *    println(mins) // 50
+     *  }}}
+     *
+     *  Here, we use the `TimeSplitter` regex within the `s` matcher, further splitting the
+     *  matched string "10.50" into its constituent parts
+     */
+    def unapplySeq(s: String) = {
```
Comment (som-snytt) [inline src/library/scala/StringContext.scala]: requisite result type here

File: src/library/scala/StringContext.scala [hunk 2]
```diff
@@ -92,7 +92,48 @@ case class StringContext(parts: String*) {
...
+      val Seq(head, tail@_*) = parts.map(java.util.regex.Pattern.quote)
+      val regex = head + tail.map("(.*)" + _).mkString
```
Comment (som-snytt) [inline src/library/scala/StringContext.scala]: Maybe one fell swoop,.

File: src/library/scala/StringContext.scala [hunk 3]
```diff
@@ -92,7 +92,48 @@ case class StringContext(parts: String*) {
    *          use of a StringBuilder.
    */
   def s(args: Any*): String = standardInterpolator(processEscapes, args, parts)
-
+  object s{
```
Comment (som-snytt) [inline src/library/scala/StringContext.scala]: The `s` needs some personal space.

**Comment by som-snytt** [comment]
A pragmatic proposal in the best sense.

One can ask if a core feature should be most "performant" because it's used all the time, or can be casual because folks who care will roll their own.

**Comment by lihaoyi** [comment]
No reason why this shouldn't be performant; I just thought it would be nice to get some feedback before spending the time implementing a linear time string-globbing algorithm from someone's blog (e.g. https://research.swtch.com/glob). If people are happy with the idea I'll happily make it fast

**Comment by mkeskells** [comment]
@lihaoyi could this be done as a macro, or at least a macro façade?
It should be more performant, but it is hard to make it a macro later as it will break binary compatibility. 
Even if we just make it a macro that calls another function for now - then we can improve the macro, to cherry-pick the cases that are improvable. I think that we will need the run-time function anyway where the format is not a literal

The binary compatibility issue is why  the `s` interpolator changes had to wait until 2.13 - https://github.com/scala/scala-dev/issues/475 and https://github.com/scala/scala/pull/6093/files worked around limitations for 2.12

**Comment by SethTisue** [comment]
this is cool, I'm in favor. anyone else on the Lightbend team want to encourage or discourage Haoyi from pursuing this further?

**Comment by lihaoyi** [comment]
yeah it shouldn't be hard to make it a macro, if people think that's the right thing to do

**Comment by smarter** [comment]
The 2.13 standard library is supposed to cross-compile between 2.13 and Dotty. Any macro addition complicates this (right now there's one macro in the standard library, the f string interpolator, which is already causing us headaches).

**Comment by lihaoyi** [comment]
I have swapped in a linear (or at least non-exponential) time algorithm for globbing, adapted from @rsc's blog. I suspect that this should be reasonably fast and good enough for most use cases; if someone wants a macro-powered NFA-generated-at-compile-time implementation to eek out the last bit of performance and be the fastest string matching algorithm on the JVM, they can implement that themselves (as I have done in FastParse for similar cases)

Notably, the current implementation doesn't really require any heavy compilation of regexes, so it's unclear how much a macro doing stuff at compile time can really help us.

I don't really know why it works, but it does. Added a moderately thorough test suite to walk through the various edge conditions in the glob matching implementation and it seems to do the right thing.

I haven't actually built/compiled/tested the scala/scala project, but have tested the code standalone, so hopefully I didn't mess up the copy-paste!

File: src/library/scala/StringContext.scala [hunk 4]
```diff
@@ -158,7 +195,90 @@ case class StringContext(parts: String*) {
...
+    var nextPatternIndex = 0
+    var nextInputIndex = 0
+
+    val matchIndices = patternChunks
+      .map(_.length + 1)
+      .inits.map(_.sum - 1)
+      .toIndexedSeq
+      .reverse
+      .drop(1)
+      .dropRight(1)
```
Comment (lihaoyi) [inline src/library/scala/StringContext.scala]: Any suggestions on how to write this bit better performance while still being reasonably elegant would be welcome

File: src/library/scala/StringContext.scala [hunk 5]
```diff
@@ -158,7 +195,90 @@ case class StringContext(parts: String*) {
...
+
+    val matchStarts = Array.fill(patternChunks.length - 1)(-1)
+    val matchEnds = Array.fill(patternChunks.length - 1)(-1)
+
+    // -1 is the magic marker number, since it cannot be a normal scala.Char
+    val pattern = patternChunks.flatMap(-1 +: _.toCharArray.map(_.toShort)).drop(1)
```
Comment (lihaoyi) [inline src/library/scala/StringContext.scala]: This bit too. We don't seem to have any `mkString` equivalent for non-string sequences

File: test/junit/scala/lang/stringinterpol/StringContextTest.scala [hunk 6]
```diff
@@ -252,4 +252,103 @@ class StringContextTest {
...
+      (StringContext.glob(Seq("", ""), "a    b")  -> Some(Seq("a    b"))),
+      (StringContext.glob(Seq("", ""), "a***b")   -> Some(Seq("a***b"))),
+      (StringContext.glob(Seq("", ""), "a")       -> Some(Seq("a"))),
+      (StringContext.glob(Seq("", ""), "b")       -> Some(Seq("b"))),
+      (StringContext.glob(Seq("", ""), "abc")     -> Some(Seq("abc"))),
+      (StringContext.glob(Seq("", ""), "cab")     -> Some(Seq("cab"))),
+      (StringContext.glob(Seq("", ""), " ab ")    -> Some(Seq(" ab "))),
+  
+      // Ambiguous lonely wildcard
+      (StringContext.glob(Seq("", "", ""), "ab")      -> Some(Seq("", "ab"))),
```
Comment (lihaoyi) [inline test/junit/scala/lang/stringinterpol/StringContextTest.scala]: Note that unlike the previous regex-based implementation, in case of ambiguity the *later* wildcards are prioritized over the earlier wildcards when characters can go to either

File: src/library/scala/StringContext.scala [hunk 7]
```diff
@@ -158,7 +195,90 @@ case class StringContext(parts: String*) {
...
+        case -1 => // do nothing
+        case n =>
+          matchStarts(n) = matchStarts(n) match{
+            case -1 => inputIndex
+            case s => math.min(s, inputIndex)
+          }
+          matchEnds(n) = matchEnds(n) match{
+            case -1 => inputIndex
+            case s => math.max(s, inputIndex)
+          }
```
Comment (lihaoyi) [inline src/library/scala/StringContext.scala:278]: For every `patternIndex` which matches one of the wildcards in `matchIndices`, we keep track of the earliest and latest parts of the input where `inputIndex` has been, and use that as the range of characters the wildcard has consumed.

File: src/library/scala/StringContext.scala [hunk 8]
```diff
@@ -158,7 +195,90 @@ case class StringContext(parts: String*) {
...
+    val matchStarts = Array.fill(patternChunks.length - 1)(-1)
+    val matchEnds = Array.fill(patternChunks.length - 1)(-1)
+
+    // -1 is the magic marker number, since it cannot be a normal scala.Char
+    val pattern = patternChunks.flatMap(-1 +: _.toCharArray.map(_.toShort)).drop(1)
+    val patternLength = pattern.length
+    val nameLength = input.length
+
+    while(patternIndex < patternLength || inputIndex < nameLength) {
+      matchIndices.indexOf(patternIndex) match{
```
Comment (lihaoyi) [inline src/library/scala/StringContext.scala]: Not sure if it's worth getting rid of this `indexOf` in favor of direct lookup, but I guess I can do that if necessary

**Comment by lihaoyi** [comment]
I've left comments on the bits of the algorithm I could use help in writing better or more elegantly. Any suggestions are welcome; feel free to just push commits to this PR if you have some improvement in mind

**Comment by lihaoyi** [comment]
Not sure what that build failure is; any ideas?

**Comment by lrytz** [comment]
I like that feature too.

Pushed a change that will fix the test failure.

**Comment by lihaoyi** [comment]
Pushed some tweaks to optimize the pre-calculation of lookup tables that we use in the main algorithm. Hopefully there shouldn't be any gross inefficiencies anymore...

**Comment by lihaoyi** [comment]
huh not sure why the test is failing in scala/scala master when it passes in scala 2.12.7; is there some easy command to run just that test?

**Comment by lihaoyi** [comment]
`sbt test/test` is giving me `[error] SERVER ERROR: GATEWAY_TIMEOUT url=https://repo.lightbend.com/typesafe/scala-sha-bootstrap/org/scala-lang/bootstrap/02fe2ed93766323a13f22c7a7e2ecdcd84259b6c/test/files/lib/annotations.jar`

**Comment by lrytz** [comment]
Our sbt build is a bit special... https://github.com/scala/scala#using-the-sbt-build. `testOnly *StringContextTest` should probably work.

**Comment by lrytz** [comment]
(or just run it in intellij)

**Comment by lihaoyi** [comment]
k fixed it, validate-main passes

**Comment by lihaoyi** [comment]
this is basically ready for review

File: src/library/scala/StringContext.scala (hunk 5 above)
Comment (ShaneDelmore) [reply src/library/scala/StringContext.scala]: I often find myself wanting something a bit more generic like intercalate. 
https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/Foldable.scala#L545

**Comment by lihaoyi** [comment]
is anyone home?

File: src/library/scala/StringContext.scala [hunk 9]
```diff
@@ -158,7 +195,127 @@ case class StringContext(parts: String*) {
...
+    * Linear time glob-matching implementation.
+    * Adapted from https://research.swtch.com/glob
+    *
+    * @param patternChunks The non-wildcard portions of the input pattern,
+    *                      separated by wildcards
+    * @param input The input you wish to match against
+    * @return None if there is no match, Some containing the sequence of matched
+    *         wildcard strings if there is a match 
+    */
+  def glob(patternChunks: Seq[String], input: String): Option[Seq[String]] = {
```
Comment (som-snytt) [inline src/library/scala/StringContext.scala:208]: Should be private, although of general utility, just to avoid committing to it as API?

File: test/junit/scala/lang/stringinterpol/StringContextTest.scala [hunk 10]
```diff
@@ -252,4 +252,103 @@ class StringContextTest {
...
+      (StringContext.glob(Seq("a", "bb", "c"), "a")         -> None),
+      (StringContext.glob(Seq("a", "bb", "c"), "abc")       -> None),
+      (StringContext.glob(Seq("a", "bb", "c"), "abb")       -> None),
+      (StringContext.glob(Seq("a", "bb", "c"), "bbc")       -> None),
+      (StringContext.glob(Seq("a", "bb", "c"), "bbaaac")    -> None),
+  
+      // Many many ambiguous wildcards
+      (StringContext.glob(Seq.fill(100)("a"), "a" * 200) -> Some(Seq.fill(98)("") :+ "a" * 100))
+    )
+    for((lhs, rhs) <- cases) assertEquals(lhs, rhs)
```
Comment (som-snytt) [inline test/junit/scala/lang/stringinterpol/StringContextTest.scala]: `assertEquals(rhs, lhs)` for the correct junit message.

File: test/junit/scala/lang/stringinterpol/StringContextTest.scala (hunk 6 above)
Comment (som-snytt) [reply test/junit/scala/lang/stringinterpol/StringContextTest.scala]: Maybe document that stars are reluctant: `val s"${a}x${b}" = "axxxb"` gives `a`, `xxb` instead of `axx`. Maybe add that unit test sample, since I think in the covering case, I can't tell which `b` is shown. Or no, L338 is a,b not ab,empty.

**Review by som-snytt — APPROVED** [review]
Clever usage! Maybe decide on code formatting before squashing? I'll take another look at the algo. There are a couple of minor artifacts like `nameLength` or `nx` where I went huh? but nbd.

**Comment by som-snytt** [comment]
I'm not actually home.

I just saw the previous invitation to push improvements to comments; maybe I'll follow up with adding some spaces, too.

**Comment by lihaoyi** [comment]
@som-snytt yeah feel free to push whatever changes you want; easier than going back and forth

**Review by SethTisue — COMMENTED** [review]
I'm in favor of merging this once @som-snytt finishes waving his magic wand over it

but I'd like another member of the Scala team to take a last look and do the final merge

**Comment by hepin1989** [comment]
I found this is very useful, thanks .will this needs a sip?

**Comment by lihaoyi** [comment]
Just to follow up here, after another two weeks. Is there anything I should do here? Or does approval mean it's out of my hands and just waiting for someone to press the button

**Comment by som-snytt** [comment]
I didn't get around to fiddling with comments, which doesn't matter because `RC` is for `re-commenting`, but they will require squashing. Seth asked someone with a bigger wand to do an incantation.

I said: "Maybe decide on code formatting before squashing?" Conditional approval is no doubt a bad practice.

**Review by lrytz — APPROVED** [review]
Rebased, pushed some tiny syntax cleanups. LGTM, thanks a lot @lihaoyi!

**Comment by lihaoyi** [comment]
Thanks lukas, sorry for being a pest!

**Comment by lrytz** [comment]
Not at all! We know that the PR queue size and delays are a little long these days, but we're actually glad when contributors push for their changes to get over the finish line :-)
